### PR TITLE
[JBPM-9331] JACCIdentityProvider must support contextUsers

### PR DESCRIPTION
**Pull request by gmunozfe**
**JIRA**: [JBPM-9331](https://issues.redhat.com/browse/JBPM-9331)

### CODE REVIEW

**Comment by elguardian** [comment]
Jenkin retest this please.

**Comment by elguardian** [comment]
Jenkins retest this please.

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/*
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java:1]: Seems right package structure folder has not been created properly.

It should be `org/jbpm/notification/listeners/` but the actual folder is just `org.jbpm.notification.listeners`

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * limitations under the License.
+*/
+
+package org.jbpm.notification.listeners;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java:22]: Please remove all these imports - not used
```suggestion
```

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import org.kie.internal.task.api.UserInfo;
+import org.kie.internal.task.api.model.NotificationEvent;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.util.Collections.singletonMap;
+
+public class CustomNotificationListener implements NotificationListener {
+
+    protected static Logger logger = LoggerFactory.getLogger(CustomNotificationListener.class);
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java:36]: this should be `final`
```suggestion
    protected static final Logger logger = LoggerFactory.getLogger(CustomNotificationListener.class);
```

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    
+    private UserTaskService userTaskService = (UserTaskService) ServiceRegistry.get().service(ServiceRegistry.USER_TASK_SERVICE);
+
+    @Override
+    public void onNotification(NotificationEvent event, UserInfo userInfo) {
+        logger.debug("onNotification with event content {}, task {} , userInfo {}",
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java:42]: `logger` expects 3 placeholders but only 2 are provided, also I think there is no harm in using `info` so we can always see the output when running tests

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+public class CustomNotificationListener implements NotificationListener {
+
+    protected static Logger logger = LoggerFactory.getLogger(CustomNotificationListener.class);
+    
+    private UserTaskService userTaskService = (UserTaskService) ServiceRegistry.get().service(ServiceRegistry.USER_TASK_SERVICE);
+
+    @Override
+    public void onNotification(NotificationEvent event, UserInfo userInfo) {
+        logger.debug("onNotification with event content {}, task {} , userInfo {}", 
+                event.getContent(), userInfo);
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java:43]: using `userInfo` will only display a memory address - maybe it's worth to display something more meaningful from `userInfo` object

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        try {
+            if ("TaskSaveContent".equals(event.getTask().getName())) {
+                Long contentId = userTaskService.saveContent(event.getTask().getId(), singletonMap("grade", "E"));
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java:46]: `contentId` is not used
```suggestion
                userTaskService.saveContent(event.getTask().getId(), singletonMap("grade", "E"));
```

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/NotificationSaveContentIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        kieContainer = KieServices.Factory.get().newKieContainer(releaseId);
+
+        createContainer(CONTAINER_ID_NOTIFICATION, releaseId);
+    }
+
+    @Test(timeout = 10000)
+    public void testSaveContentOnNotification() throws Exception {
+        Long processInstanceId = processClient.startProcess(CONTAINER_ID_NOTIFICATION, PROCESS_ID_NOTIFICATION);
+        assertNotNull(processInstanceId);
+        assertTrue(processInstanceId.longValue() > 0);
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/NotificationSaveContentIntegrationTest.java:54]: Unnecessary unboxing
```suggestion
        assertTrue(processInstanceId > 0);
```

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/NotificationSaveContentIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+        try {
+            List<TaskSummary> tasks=taskClient.findTasksAssignedAsPotentialOwner(USER_YODA, 0, 10);
+            assertEquals(1, tasks.size());
+            
+            //Task is claimed during onNotification, as well as saving output content
+            KieServerSynchronization.waitForTaskStatus(taskClient, tasks.get(0).getId(), "Reserved");
+            
+            tasks=taskClient.findTasksByStatusByProcessInstanceId(processInstanceId, Arrays.asList("Reserved"), 0, 10);
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/NotificationSaveContentIntegrationTest.java:63]:
```suggestion
            tasks=taskClient.findTasksByStatusByProcessInstanceId(processInstanceId, Collections.singletonList("Reserved"), 0, 10);
```

**Review by afalhambra — CHANGES_REQUESTED** [review]
Awesome work @gmunozfe! Just a few comments, most of them only cosmetic changes. Only the package folder structure seems not correct for CustomNotificationListener class.

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java (hunk 2 above)
Comment (gmunozfe) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java]: done

File: kie-server-parent/kie-server-services/kie-server-services-common/src/test/java/org/kie/server/services/impl/security/JACCIdentityProviderWildFlyTomcatTest.java [hunk 9]
```diff
@@ -107,6 +108,19 @@ public void testGetRoles() throws Exception {
         assertFalse(jaccIdentityProvider.hasRole("non_existing_role"));
         assertFalse(jaccIdentityProvider.hasRole(null));
     }
+    
+    @Test
+    public void testContextUsers() throws Exception {
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-server-parent/kie-server-services/kie-server-services-common/src/test/java/org/kie/server/services/impl/security/JACCIdentityProviderWildFlyTomcatTest.java]: done

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/pom.xml [hunk 10]
```diff
@@ -144,6 +148,26 @@
...
+        <plugin>
+          <groupId>org.apache.maven.plugins</groupId>
+          <artifactId>maven-antrun-plugin</artifactId>
+          <executions>
+            <execution>
+              <id>prepare-kie-server</id>
+              <phase>pre-integration-test</phase>
+              <configuration>
+                <target>
+                  <unzip src="${maven.dependency.org.kie.server.kie-server.ee7.war.path}" dest="${project.build.directory}/unzipped-kie-server"/>
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/pom.xml]: you're right, switched to ee8

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java (hunk 1 above)
Comment (gmunozfe) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java]: my bad, fixed

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java (hunk 3 above)
Comment (gmunozfe) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java]: done

**Review by afalhambra — APPROVED** [review]
Very well done @gmunozfe!

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java (hunk 4 above)
Comment (gmunozfe) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java]: just kept content, as it's the only one meaningful, right

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java (hunk 6 above)
Comment (gmunozfe) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/main/java/org.jbpm.notification.listeners/CustomNotificationListener.java]: done

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/NotificationSaveContentIntegrationTest.java (hunk 7 above)
Comment (gmunozfe) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/NotificationSaveContentIntegrationTest.java]: done

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/NotificationSaveContentIntegrationTest.java (hunk 8 above)
Comment (gmunozfe) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/NotificationSaveContentIntegrationTest.java]: done

**Review by elguardian — APPROVED** [review]
LGTM. waiting for CI clean

**Comment by sutaakar** [comment]
jenkins execute springboot build

**Comment by gmunozfe** [comment]
jenkins execute springboot build

**Comment by gmunozfe** [comment]
jenkins execute springboot build

**Comment by gmunozfe** [comment]
jenkins execute springboot build

**Comment by gmunozfe** [comment]
jenkins execute springboot build

**Comment by gmunozfe** [comment]
jenkins execute springboot build

**Comment by gmunozfe** [comment]
jenkins retest this

**Comment by afalhambra** [comment]
jenkins retest this

**Pull request merged**